### PR TITLE
Create subdirs when creating new note

### PR DESCRIPTION
**Pull request by Sleepful**
## Proposed change

Context inside commit messages and linked issues.

<details>

```
Author: Sleepful <josepablov@gmail.com>
Date:   Thu Jul 6 14:21:23 2023 -0600

    Update check_dir_and_ask: async for vim.ui.select

    vim.ui.select is meant to be used as an asynchronous operation,
    in order to do this, it receives a callback as a third argument.

    This commit adjusts check_dir_and_ask to leverage the asynchronous
    usage of vim.ui.select, as a side effect to make this work, I also
    refactored the code that depends on the function in order to leverage
    the asynchronous operations.

    References:
    - https://github.com/stevearc/dressing.nvim/issues/80#issuecomment-1380734202
    - https://github.com/stevearc/dressing.nvim
    - https://github.com/neovim/neovim/pull/15771

    Fixes:
    - https://github.com/renerocksai/telekasten.nvim/issues/266

    One reason this bug can sneak into the codebase is because the
    default unmodified behavior of vim.ui.select is to use the
    :Messages section of neovim to ask for input and this UI is
    a synchronous blocking operation. However, this is not the
    intended usage of vim.ui.select as presented above.

commit b6ff784fbb46f170c8851be696a864e101606be5
Author: Sleepful <josepablov@gmail.com>
Date:   Thu Jul 6 14:20:47 2023 -0600

    Update README with correct configuration name

commit f2f517d0c80f1a89629bf68fc550e53070fc8da8
Author: Sleepful <josepablov@gmail.com>
Date:   Wed Jul 5 22:34:48 2023 -0600

    Create subdirs when creating new note

    previously it failed with error message


```

</details>

## Replicate

One way to look at the broken functionality before this commit is to install `dressing.nvim` along with `Telekasten` and try to trigger `check_dir` function

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [x] Code quality improvements to existing code or addition of tests
- [x] Documentation update

It is difficult to split this into separate PRs

## Additional information

- This PR fixes or closes issue: 
- fixes https://github.com/renerocksai/telekasten.nvim/issues/266
- fixes https://github.com/renerocksai/telekasten.nvim/issues/217
- same as above, but FR: https://github.com/renerocksai/telekasten.nvim/issues/205
- This PR is related to issue:
- Might be related to other issues, maybe not.

## Checklist

- [x] I am running the **latest** version of the plugin.
- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] The code has been formatted using Stylua (a `.stylua.toml` file is provided)
- [ ] The code has been checked with luacheck (a `.luacheckrc` file is provided)
- [ ] The `README.md` has been updated according to this change.
- [ ] The `doc/telekasten.txt` helpfile has been updated according to this change.

### CODE REVIEW

File: lua/telekasten.lua [hunk 1]
```diff
@@ -1975,428 +2069,444 @@ local function FollowLink(opts)
...
+                        local uuid = fileutils.new_uuid(uuid_type)
+                        create_note_from_template(
+                            title,
+                            uuid,
+                            pinfo.filepath,
+                            pinfo.template,
+                            pinfo.calendar_info,
+                            function()
+                                opts.erase = true
+                                opts.erase_file = fname
```
Comment (lambtho12) [inline lua/telekasten.lua:2198]: Stylua complains here that fname is not defined properly. Why not keep `pinfo.filepath`?

**Review by lambtho12 — CHANGES_REQUESTED** [review]

**Comment by lambtho12** [comment]
I have not reviewed the whole PR, but some preliminary tests on my side seems to indicate it works well. I activated the pipeline so any further commits will be checked directly for linting issues.

It is a rather large PR, I will try to review it by the end of the week-end and merge it for Monday.

Anyway, thank you already for this massive contribution!

**Comment by Sleepful** [comment]
No problem, sincerely loving this plugin.

I will add comments to the diff tomorrow, in order to make it easier to parse for you.

File: lua/telekasten.lua [hunk 2]
```diff
@@ -183,7 +183,7 @@ local function generate_note_filename(uuid, title)
     end
 end
 
-local function check_dir_and_ask(dir, purpose)
+local function check_dir_and_ask(dir, purpose, callback)
```
Comment (Sleepful) [inline lua/telekasten.lua:186]: this is the biggest source of change. Added `callback` param to `check_dir_and_ask`, now all callsites to `check_dir_and_ask` need to pass the "after selection" operations as a callback.

File: lua/telekasten.lua [hunk 3]
```diff
@@ -201,34 +201,44 @@ local function check_dir_and_ask(dir, purpose)
                     vim.cmd('echomsg " "')
                     vim.cmd('echomsg "' .. dir .. ' created"')
                     ret = true
+                    callback(ret)
```
Comment (Sleepful) [inline lua/telekasten.lua:204]: "resumes" the code that was passed as `callback`

File: lua/telekasten.lua [hunk 4]
```diff
@@ -201,34 +201,44 @@ local function check_dir_and_ask(dir, purpose)
...
                 else
                     -- unreachable: plenary.Path:mkdir() will error out
                     tkutils.print_error("Could not create directory " .. dir)
                     ret = false
+                    callback(ret)
```
Comment (Sleepful) [inline lua/telekasten.lua:209]: "resumes" the code that was passed as `callback`

File: lua/telekasten.lua [hunk 5]
```diff
@@ -201,34 +201,44 @@ local function check_dir_and_ask(dir, purpose)
...
                 end
             end
         end)
     else
         ret = true
+        if callback ~= nil then
+            callback(ret)
```
Comment (Sleepful) [inline lua/telekasten.lua:216]: even though `vim.ui.select` was not called in this branch, code still needs to continue through callback

File: lua/telekasten.lua [hunk 6]
```diff
@@ -201,34 +201,44 @@ local function check_dir_and_ask(dir, purpose)
...
+        end
+        return ret
     end
-    return ret
 end
 
-local function global_dir_check()
+local function global_dir_check(callback)
```
Comment (Sleepful) [inline lua/telekasten.lua:222]: this is one of the main functions affected by the refactor of `check_dir_and_ask`, therefore all functions that call `global_dir_check` are affected and need to pass in a callback

File: lua/telekasten.lua [hunk 7]
```diff
@@ -245,36 +255,38 @@ local function make_config_path_absolute(path)
 end
 
 local function recursive_substitution(dir, old, new)
-    if not global_dir_check() then
-        return
-    end
+    global_dir_check(function(dir_check)
```
Comment (Sleepful) [inline lua/telekasten.lua:258]: so most changes look like this, code is the same but now it includes a `function` callback

File: lua/telekasten.lua (hunk 1 above)
Comment (Sleepful) [reply lua/telekasten.lua]: hm, looks like this was an oversight on my part

Comment (Sleepful) [reply lua/telekasten.lua]: updated

File: lua/telekasten.lua [hunk 8]
```diff
@@ -448,21 +463,30 @@ local function create_note_from_template(
...
-            templates.subst_templated_values(
-                line,
-                title,
-                calendar_info,
-                uuid,
-                M.Cfg.calendar_opts.calendar_monday
-            ) .. "\n"
-        )
-    end
+    local file_dir = filepath:match("(.*/)") or ""
```
Comment (Sleepful) [inline lua/telekasten.lua:466]: get directory from filepath

File: lua/telekasten.lua [hunk 9]
```diff
@@ -448,21 +463,30 @@ local function create_note_from_template(
...
+            ofile:write(
+                templates.subst_templated_values(
+                    line,
+                    title,
+                    calendar_info,
+                    uuid,
+                    M.Cfg.calendar_opts.calendar_monday
+                ) .. "\n"
+            )
+        end
```
Comment (Sleepful) [inline lua/telekasten.lua:484]: this is the bit that asks to create non-existent directories from the note title

**Review by lambtho12 — APPROVED** [review]

**Comment by lambtho12** [comment]
Thank you very much for this PR! Feel free to tackle other issues or pickup the big refactoring I started on the `refact` branch :smile:

**Pull request merged**